pkg/importer/data_processor.go

```diff
@@ -0,0 +1,211 @@
+limitations under the License.
+*/
+
+package importer
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog"
+	"kubevirt.io/containerized-data-importer/pkg/image"
+	"kubevirt.io/containerized-data-importer/pkg/util"
+)
+
+var qemuOperations = image.NewQEMUOperations()
+
+// ProcessingPhase is the current phase being processed.
+type ProcessingPhase string
+
+const (
+	// Info is the first phase, durlng this phase the provider obtains information needed to determine which phase to go to next.
+	Info ProcessingPhase = "Info"
+	// Transfer is the phase in which the data provider writes data to the scratch space.
+	Transfer ProcessingPhase = "Transfer"
+	// TransferTarget is the phase in which the data provider writes data directly to the target space without conversion.
+	TransferTarget ProcessingPhase = "TransferTarget"
+	// Process is the phase in which the data provider processes the data just written to the scratch space.
+	Process ProcessingPhase = "Process"
+	// Convert is the phase in which the data is taken from the url provided by the provider, and it is converted to the target RAW disk image format.
+	// The url can be an http end point or file system end point.
+	Convert ProcessingPhase = "Convert"
+	// Resize the disk image, this is only needed when the target contains a file system (block device do not need a resize)
+	Resize ProcessingPhase = "Resize"
+	// Complete is the phase where the entire process completed successfully and we can exit gracefully.
+	Complete ProcessingPhase = "Complete"
+	// Error is the phase in which we encountered an error and need to exit ungracefully.
+	Error ProcessingPhase = "Error"
+)
+
+// ErrRequiresScratchSpace indicates that we require scratch space.
+var ErrRequiresScratchSpace = fmt.Errorf("Scratch space required and none found")
+
+// DataProviderInterface is the interface all data providers should implement.
+type DataProviderInterface interface {
+	// Info() is called to get initial information about the data.
+	Info() (ProcessingPhase, error)
+	// Transfer() is called to transfer the data from the source to a temporary location.
+	Transfer(path string) (ProcessingPhase, error)
+	// Process() is called to do any special processing before giving the url to the data back to the processor
+	Process() (ProcessingPhase, error)
+	// Geturl returns the url that the data processor can use when converting the data.
+	GetURL() *url.URL
+	// Close closes any readers or other open resources.
+	Close() error
+}
+
+// DataProcessor holds the fields needed to process data from a data provider.
+type DataProcessor struct {
+	// currentPhase is the phase the processing is in currently.
+	currentPhase ProcessingPhase
+	// provider provides the data for processing.
+	provider DataProviderInterface
+	// destination file. will be DataDir/disk.img if file system, or a block device (if a block device, then DataDir will not exist).
+	dest string
```

Comment:
may be dataFile instead dest?

Reply:
That would be consistent with the terminology we are using elsewhere, agreed will change.